phoenix-core/src/main/java/org/apache/phoenix/expression/BaseJSONExpression.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.phoenix.expression;
+
+import java.util.List;
+
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.phoenix.expression.visitor.ExpressionVisitor;
+import org.apache.phoenix.schema.tuple.Tuple;
+import org.apache.phoenix.schema.types.PDataType;
+
+public abstract class BaseJSONExpression extends BaseCompoundExpression {
+	public BaseJSONExpression(List<Expression> children) {
+        super(children);
+    }
+	public BaseJSONExpression() {
+    }
+	public abstract boolean evaluate(Tuple tuple, ImmutableBytesWritable ptr);
+	public abstract <T> T accept(ExpressionVisitor<T> visitor);
+	public abstract PDataType getDataType();
+	public PDataType getRealDataType(){
+		return null;
```

Comment:
Because the method -- getDataType() is used to check before evaluate the operation , but JSON data type is hybrid data format. So if we haven't watched the value of specific key in JSON , we cannot know the type of JSON value. 
For example : 
the following operation : 
```
'{"a":3 , "b":"string"}' -> 'a'
```
this type will be as integer
but another operation :
```
'{"a":"string","b":2.3}' -> 'a'
```
this type will be as varchar

So we use method -- getRealDataType() to ensure the type after we evaluate the operation